src/SplitTuples.cpp

```diff
@@ -165,31 +155,173 @@ class SplitTuples : public IRMutator {
+        // things in a mutex. We express this using separate atomic
+        // nodes per store. If there's no mutex involved at all, then
+        // there's no benefit in packing things together into a single
+        // critical section.
+        bool separate_atomic_nodes_per_store =
+            ((atomic && atomic->mutex_name.empty()) ||
+             (clusters.size() == op->values.size()));
+
+        // For each cluster, build a list of scalar provide
+        // statements, and a list of lets to wrap them.
+        vector<Stmt> result;
+        for (auto &c : clusters) {
+            if (c.empty()) {
+                continue;
+            }
+            std::sort(c.begin(), c.end());
+            vector<Stmt> provides;
+            vector<pair<string, Expr>> lets;
+
+            Stmt s;
+
+            if (c.size() == 1) {
+                // Just make a provide node
+                int i = *c.begin();
+                string name = op->name + "." + std::to_string(i);
+                s = Provide::make(name, {mutate(op->values[i])}, args);
+            } else {
+                // Make a list of let statements that compute the
+                // values (doing any loads), and then a block of
+                // provide statements that do the stores.
+                for (auto i : c) {
+                    string name = op->name + "." + std::to_string(i);
+                    string var_name = name + ".value";
+                    Expr val = mutate(op->values[i]);
+                    if (!is_undef(val)) {
+                        lets.emplace_back(var_name, val);
+                        val = Variable::make(val.type(), var_name);
+                    }
+                    provides.push_back(Provide::make(name, {val}, args));
+                }
+
+                s = Block::make(provides);
+
+                while (!lets.empty()) {
+                    auto p = lets.back();
+                    lets.pop_back();
+                    s = LetStmt::make(p.first, p.second, s);
+                }
+            }
 
-        while (!lets.empty()) {
-            auto p = lets.back();
-            lets.pop_back();
-            result = LetStmt::make(p.first, p.second, result);
+            if (atomic && separate_atomic_nodes_per_store) {
+                s = Atomic::make(atomic->producer_name, atomic->mutex_name, s);
+            }
+
+            internal_assert(s.defined());
+            result.push_back(s);
+        }
+
+        {
+            Stmt s = Block::make(result);
+            if (atomic && !separate_atomic_nodes_per_store) {
+                s = Atomic::make(atomic->producer_name, atomic->mutex_name, s);
+            }
+            return s;
         }
+    }
 
-        return result;
+    Stmt visit(const Provide *op) override {
+        return visit_provide(op, nullptr);
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (const Provide *p = op->body.as<Provide>()) {
```

Comment:
Is it possible to have other nodes between the Atomic node and the Provide node? Also can we have multiple Provides inside an Atomic node?

Reply:
Not at this point in lowering. I added a comment to that effect.